go/tls/tls.go

```diff
@@ -0,0 +1,682 @@
+//	}
+//
+// TLS variants (RFC 5246 s4.6.1) are only supported when the value of the
+// associated enumeration type is available earlier in the same enclosing
+// struct and is marked with the selector tag.
+//
+// For example, a TLS structure:
+//
+//   enum { e1(1), e2(2) } EnumType;
+//   struct {
+//      EnumType sel;
+//      select(sel) {
+//         case e1: uint16
+//         case e2: uint32
+//      } data;
+//   } VariantItem;
+//
+// would have a corresponding Go type:
+//
+//   type VariantItem struct {
+//      Sel    tls.Enum  `tls:"maxval:2"`
+//      Data16 *uint16   `tls:"selector:Sel,val:1"`
+//      Data32 *uint32   `tls:"selector:Sel,val:2"`
+//    }
+//
+// TLS fixed-length vectors of types other than opaque or uint8 are not supported.
+//
+// For TLS variable-length vectors that are themselves used in other vectors,
+// create a single-field structure to represent the inner type. For example, for
+//
+//   opaque InnerType<1..65535>;
+//   struct {
+//     InnerType inners<1,65535>;
+//   } Something;
+//
+// convert to:
+//
+//   type InnerType struct {
+//      Val    []byte       `tls:"minlen:1,maxlen:65535"`
+//   }
+//   type Something struct {
+//      Inners []InnerType  `tls:"minlen:1,maxlen:65535"`
+//   }
+//
+// If the encoded value does not fit in the Go type, Unmarshal returns a parse error.
+func Unmarshal(b []byte, val interface{}) (rest []byte, err error) {
+	return UnmarshalWithParams(b, val, "")
+}
+
+// UnmarshalWithParams allows field parameters to be specified for the
+// top-level element. The form of the params is the same as the field tags.
+func UnmarshalWithParams(b []byte, val interface{}, params string) (rest []byte, err error) {
+	v := reflect.ValueOf(val).Elem()
+	info, err := fieldTagToFieldInfo(params, "")
+	if err != nil {
+		return nil, err
+	}
+	offset, err := parseField(v, b, 0, info)
+	if err != nil {
+		return nil, err
+	}
+	return b[offset:], nil
+}
+
+// Return the number of bytes needed to encode values up to (and including) x.
+func byteCount(x uint64) uint {
+	switch {
+	case x < 0x100:
+		return 1
+	case x < 0x10000:
+		return 2
+	case x < 0x1000000:
+		return 3
+	case x < 0x100000000:
+		return 4
+	case x < 0x10000000000:
+		return 5
+	case x < 0x1000000000000:
+		return 6
+	case x < 0x100000000000000:
+		return 7
+	default:
+		return 8
+	}
+}
+
+type fieldInfo struct {
+	count    uint // Number of bytes
+	countSet bool
+	minlen   uint64 // Only relevant for slices
+	maxlen   uint64 // Only relevant for slices
+	selector string // Only relevant for select sub-values
+	val      uint64 // Only relevant for select sub-values
+	name     string // Used for better error messages
+}
+
+// Given a tag string, return a fieldInfo describing the field.
+func fieldTagToFieldInfo(str string, name string) (info *fieldInfo, err error) {
+	for _, part := range strings.Split(str, ",") {
+		switch {
+		case strings.HasPrefix(part, "maxval:"):
+			v, err := strconv.ParseUint(part[7:], 10, 64)
```

Comment:
It kind of was WAI, in that a malformed clause just gets silently ignored as if it wasn't there; added a comment to that effect.